### PR TITLE
Aug: Add simple support to dict for AugSequential

### LINKED ISSUE #2119: AugmentationSequential: accept sample dict as input
## 🚀 Feature

I would like `AugmentationSequential` to support a dictionary as input.

## Motivation

I'm a [TorchGeo](https://github.com/microsoft/torchgeo) developer. In TorchGeo, every dataset returns a sample dictionary like so:
```python
sample = {
    "input": torch.tensor(...),
    "mask": torch.tensor(...),
    "bbox": torch.tensor(...),
    ...
}
```
(the exact key names don't match at the moment, but we're working on standardizing those)

## Pitch

With the feature I'm envisioning, the following would work:
```python
augs = AugmentationSequential(...)
sample = augs(sample)
```

The exact implementation details would still need to be worked out, but `*args` would go from `Tensor` to `Union[Tensor, Dict[str, Tensor]]`. The dictionary may contain keys that Kornia doesn't know how to support, and these should be ignored. If a sample dictionary contains a known key that the user doesn't want to transform, they can simply pass `data_keys` to override the default detection. If the input is a dict, the output should also be a dict. If implemented correctly, this feature will be backwards compatible with the old behavior so people can still pass these inputs in manually if they want to.

## Alternatives

At the moment, to use Kornia augmentations, we have to use:
```python
augs = AugmentationSequential(..., data_keys=["input", "mask", "bbox", ...])
sample["input"], sample["mask"], sample["bbox"], ... = augs(sample["input"], sample["mask"], sample["bbox"], ...)
```
As you can see, this is much more verbose than necessary. There's no reason we need to duplicate the list of keys so many times.

## Additional context

If this is something you would be interested in, I would be happy to submit a PR to support this. Just wanted to gauge interest first before working on it.

@isaaccorley

### PR DESCRIPTION
**Pull request by johnnv1**
#### Changes
With this patch, we will be able to use the `AugmentationSequential` within a dictionary as input instead of `datakeys`. Just an extra option of input for the container. In other PR(s) in the future, we will look at how to improve this functionality to better handle cases instance segmentation 😄 

Fixes #2119

Instead of
```python
...
aug = AugmentationSequential(..., data_keys=["input", "mask", "bbox", "keypoints"])
o_img, o_mask, o_bbox, 0_keypoints = aug(input, mask, bbox, keypoints)
```

we will be able to use 

```python
...
aug = AugmentationSequential(..., data_keys=None)
data = {'input': input, 'mask': mask, 'bbox': bbox, 'keypoints': keypoints}

augmented_data = aug(data)
print(augmented_data['input'].shape)
print(augmented_data['mask'].shape)
print(augmented_data['bbox'].shape)
print(augmented_data['keypoints'].shape)
```

To use duplicated/multiple data keys equals in your dictionary,  you need to match the datakey name in the start of the dictionary key.
```python
...
aug = AugmentationSequential(..., data_keys=None)
data = {'input': input, 'mask-a': mask_a, 'mask_b': mask_b, 'bbox': bbox}

augmented_data = aug(data)
print(augmented_data['input'].shape)
print(augmented_data['mask-a'].shape)
print(augmented_data['mask_b'].shape)
print(augmented_data['bbox'].shape)
```

#### Type of change

- [x] 📚  Documentation Update
- [x] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

cc @adamjstewart

### CODE REVIEW

File: kornia/augmentation/container/augment.py [hunk 1]
```diff
@@ -364,11 +383,35 @@ def forward(  # type: ignore[override]
...
+        if isinstance(original_keys, list):
+            return {k: v for v, k in zip(outputs, original_keys)}
+
         if len(outputs) == 1 and isinstance(outputs, list):
             return outputs[0]
 
         return outputs
 
+    def _read_datakeys_from_dict(self, data: dict[str, Tensor]) -> List[DataKey]:
+        # TODO: Support multiple dict levels
```
Comment (johnnv1) [inline kornia/augmentation/container/augment.py:395]: add tests, in theory the loop bellow already do it

**Comment by johnnv1** [comment]
@shijianjian to be able to use 

```
data = {
    'input-a': ...,
    'input-b': ...,
    'mask-a': ...,
    'mask-b': ...,
}

```
we should first merge #2740 ? or do you see any other way than transforming it into a list?

File: kornia/augmentation/container/augment.py (hunk 1 above)
Comment (johnnv1) [reply kornia/augmentation/container/augment.py:395]: + tests with random things after the datakey name

**Comment by adamjstewart** [comment]
Excited to try this out! Unfortunately I'm booked solid for the rest of the month, so don't wait on me if you think it's ready to merge. Last time I tried to make my own PR, I got somewhere around this far, but every step I took resulted in cascading changes across the entire library due to type hints and things. But I trust you a lot more than myself to try to get this working.

**Comment by shijianjian** [comment]
> @shijianjian to be able to use
> 
> ```
> data = {
>     'input-a': ...,
>     'input-b': ...,
>     'mask-a': ...,
>     'mask-b': ...,
> }
> ```
> 
> we should first merge #2740 ? or do you see any other way than transforming it into a list?

Yes. We can merge #2740 if the previous bug has been fixed.

File: kornia/augmentation/container/augment.py [hunk 2]
```diff
@@ -241,15 +271,23 @@ def identity_matrix(self, input: Tensor) -> Tensor:
...
         Number of input tensors must align with the number of``data_keys``. If ``data_keys`` is not set, use
         ``self.data_keys`` by default.
         """
+        original_keys = None
+        if len(args) == 1 and isinstance(args[0], dict):
+            original_keys, data_keys, args = self._preproc_dict_data(args[0])
+
+        # args here should already be `DataType`
+        # NOTE: how to right type to: unpacked args <-> tuple of args to unpack
+        # issue with `self._preproc_dict_data` return args type
```
Comment (johnnv1) [inline kornia/augmentation/container/augment.py:289]: how can i type this in the right way?

```
kornia/augmentation/container/augment.py:291: error: Argument 1 to "_validate_args_datakeys" of
"AugmentationSequential" has incompatible type
"*tuple[Tensor | list[Tensor] | Boxes | Keypoints | dict[str, Tensor | list[Tensor] | Boxes | Keypoints], ...]";
expected "Tensor | list[Tensor] | Boxes | Keypoints"  [arg-type]
            self._validate_args_datakeys(*args, data_keys=self.transform_op.data_keys)
```

**Comment by johnnv1** [comment]
Using our panda asset as example:

```python
import torch
import kornia
import kornia.augmentation as K

img = kornia.io.load_image('./panda.jpg', kornia.io.ImageLoadType.RGB32)

aug_dict = K.AugmentationSequential(
    K.ColorJiggle(0.1, 0.1, 0.1, 0.1, p=1.0),
    K.RandomHorizontalFlip(p=1.0),
    K.ImageSequential(K.RandomHorizontalFlip(p=1.0)),
    K.ImageSequential(K.ColorJiggle(0.1, 0.1, 0.1, 0.1, p=1.0)),
    data_keys=None,
    same_on_batch=False,
    random_apply=10,
)

# Some data annotations example
h, w = img.shape[-2:]

boxes_head = torch.tensor([[[355, 10, 305, 240]]]) # xywh
keypoints = torch.tensor([[[465, 115], [545, 116]]])
mask_head = torch.zeros(1, h, w)
mask_head[..., 10:250, 355:660] = 1

mask_nodata = torch.zeros(h, w)
mask_nodata[..., :, 0:150] = 1

# Handle it to proper shapes/structures for AugSeq
from kornia.geometry.boxes import Boxes
from kornia.geometry.keypoints import Keypoints

data = {
    'image': img, # 3xHxW
    'bbox': Boxes.from_tensor(boxes_head, mode="xywh"), # 1x1x4 #xywh
    'keypoints': Keypoints.from_tensor(keypoints.float()), # 1x2x2
    'mask': mask_head[None, ...], # 1xHxW -> # 1x1xHxW
    'mask-nodata': mask_nodata[None, None, ...] # HxW-> # 1x1xHxW
}

# Augment data
augmented_data = aug_dict(data)

# Invert back the data
inverse_data = aug_dict.inverse(augmented_data)
```

**data**: image
**augmented_data**: 
image
**inverse_data**:
image

**Comment by johnnv1** [comment]
@shijianjian The API looks to just work having N equals data keys, not showing the need to pack this to be treated as a list... can you review if I'm missing something here?

**Comment by shijianjian** [comment]
> @shijianjian The API looks to just work having N equals data keys

This is expected I think.

> not showing the need to pack this to be treated as a list... can you review if I'm missing something here?

You meant to wrap `{'mask-a': ..., 'mask-b': ...}` as a list? We need to support lists for `mask-a` and `mask-b` separately later. So that for a batch size 2 data, each of them can have different number of masks for different number of instances.

File: kornia/augmentation/container/augment.py (hunk 2 above)
Comment (johnnv1) [reply kornia/augmentation/container/augment.py:289]: i do not know how to properly type this one, can I go with the type ignore here?

File: kornia/augmentation/container/augment.py [hunk 3]
```diff
@@ -168,12 +168,37 @@ class AugmentationSequential(TransformMatrixMinIn, ImageSequential):
...
+        >>> out = aug_dict(data)
+        >>> out.keys()
+        dict_keys(['image', 'mask', 'mask-b', 'bbox', 'bbox-other'])
     """
 
     def __init__(
         self,
         *args: Union[_AugmentationBase, ImageSequential],
-        data_keys: Union[List[str], List[int], List[DataKey]] = [DataKey.INPUT],
+        data_keys: Optional[Union[List[str], List[int], List[DataKey]]] = [DataKey.INPUT],
```
Comment (edgarriba) [inline kornia/augmentation/container/augment.py:201]: should we just deprecate this option (via warning) ?

File: kornia/augmentation/container/augment.py [hunk 4]
```diff
@@ -168,12 +168,37 @@ class AugmentationSequential(TransformMatrixMinIn, ImageSequential):
...
+        >>> aug_dict = K.AugmentationSequential(
+        ...    K.ColorJiggle(0.1, 0.1, 0.1, 0.1, p=1.0),
+        ...    K.RandomHorizontalFlip(p=1.0),
+        ...    K.ImageSequential(K.RandomHorizontalFlip(p=1.0)),
+        ...    K.ImageSequential(K.ColorJiggle(0.1, 0.1, 0.1, 0.1, p=1.0)),
+        ...    data_keys=None,
+        ...    same_on_batch=False,
+        ...    random_apply=10,
+        ... )
+        >>> data = {'image': img, 'mask': mask[0], 'mask-b': mask[1], 'bbox': bbox[0], 'bbox-other':bbox[1]}
```
Comment (edgarriba) [inline kornia/augmentation/container/augment.py:192]: can you update the docs / tutorials (later) stating how the keywords names work ? it handled properly to detect uniques keywords ?

File: kornia/augmentation/container/augment.py (hunk 3 above)
Comment (johnnv1) [reply kornia/augmentation/container/augment.py:201]: i think it is an extra option than a replacement for data_keys, at least for now...

File: kornia/augmentation/container/augment.py [hunk 5]
```diff
@@ -241,20 +271,28 @@ def identity_matrix(self, input: Tensor) -> Tensor:
...
+            original_keys, data_keys, args = self._preproc_dict_data(args[0])
+
+        # args here should already be `DataType`
+        # NOTE: how to right type to: unpacked args <-> tuple of args to unpack
+        # issue with `self._preproc_dict_data` return args type
+
         self.transform_op.data_keys = self.transform_op.preproc_datakeys(data_keys)
 
-        self._validate_args_datakeys(*args, data_keys=self.transform_op.data_keys)
+        self._validate_args_datakeys(*args, data_keys=self.transform_op.data_keys)  # typing: ignore
```
Comment (johnnv1) [inline kornia/augmentation/container/augment.py:293]:
```suggestion
        self._validate_args_datakeys(*args, data_keys=self.transform_op.data_keys)  # type: ignore
```

File: kornia/augmentation/container/augment.py [hunk 6]
```diff
@@ -241,20 +271,28 @@ def identity_matrix(self, input: Tensor) -> Tensor:
...
 
-        in_args = self._arguments_preproc(*args, data_keys=self.transform_op.data_keys)
+        in_args = self._arguments_preproc(*args, data_keys=self.transform_op.data_keys)  # typing: ignore
```
Comment (johnnv1) [inline kornia/augmentation/container/augment.py:295]:
```suggestion
        in_args = self._arguments_preproc(*args, data_keys=self.transform_op.data_keys)  # type: ignore
```

File: kornia/augmentation/container/augment.py [hunk 7]
```diff
@@ -341,18 +382,23 @@ def _arguments_postproc(
...
 
+        # Unpack/handle dictionary args
+        original_keys = None
+        if len(args) == 1 and isinstance(args[0], dict):
+            original_keys, data_keys, args = self._preproc_dict_data(args[0])
+
         self.transform_op.data_keys = self.transform_op.preproc_datakeys(data_keys)
 
-        self._validate_args_datakeys(*args, data_keys=self.transform_op.data_keys)
+        self._validate_args_datakeys(*args, data_keys=self.transform_op.data_keys)  # typing: ignore
```
Comment (johnnv1) [inline kornia/augmentation/container/augment.py:399]:
```suggestion
        self._validate_args_datakeys(*args, data_keys=self.transform_op.data_keys)  # type: ignore
```

File: kornia/augmentation/container/augment.py [hunk 8]
```diff
@@ -341,18 +382,23 @@ def _arguments_postproc(
...
 
-        in_args = self._arguments_preproc(*args, data_keys=self.transform_op.data_keys)
+        in_args = self._arguments_preproc(*args, data_keys=self.transform_op.data_keys)  # typing: ignore
```
Comment (johnnv1) [inline kornia/augmentation/container/augment.py:401]:
```suggestion
        in_args = self._arguments_preproc(*args, data_keys=self.transform_op.data_keys)  # type: ignore
```

**Comment by johnnv1** [comment]
@edgarriba @shijianjian i think this one is fine to go

**Review by edgarriba — APPROVED** [review]
